### DISCUSSION TITLE
Does memory_limit configuration parameter apply at system/database level or process level?

### DISCUSSION THREAD

**Discussion by parvizfarnia** [duckdb/duckdb#7811] [Q&A] [answered]
Hello everyone,

I would like to ask a question about the `memory_limit` configuration option. I work most of the time in a datalab which is a shared environment among many users. Obviously, resources (CPU and RAM) are limited. That's why I always use the following two options in every program using `DuckDB` to be sure that I'm not going to bypass the attributed quota in terms of available resources by accident:

```
SET memory_limit='8GB';
SET temp_directory='/usr/myuser/duckdb/tmp'
```

And so far, this has always fulfilled my requirement.

Currently, I am working on a `Python` program that will create 8 processes via `concurrent.futures.ProcessPoolExecutor`. There is no data race as all the created processes access the database just for reading data (`SELECT`). Basically each of these processes will call a function in which a connection to my `DuckDB` database will be created as following:

```
con = duckdb.connect(
    database='my-db.duckdb', 
    read_only=True
)

con.execute("SET memory_limit='8GB'")
con.execute("SET temp_directory='/usr/myuser/duckdb/tmp'")
```

Now my concern is bout the `memory_limit` parameter. Given that, each process calls my function which includes

 ```con.execute("SET memory_limit='8GB'")```

And given that there 8 processes, does this mean that the program can potentially consume up to `8 * 8 = 64 GB` of RAM or the value of this parameter, that is `8 GB`, is shared at system/database level among all processes even when each process, executes the `SET memory_limit` statement once?

Thanks in advance.

**Comment by parvizfarnia** [comment]
Any idea?

**Reply by Tishj** [reply, marked as answer]
If each process creates a duckdb connection and sets the limit to 8GB that means they could potentially consume up to `8 * 8 = 64 GB` of RAM.

DuckDB is in essence a single process database, without any third-party management every process is standalone.
Database settings are not persistent either, I don't know if they are planned to be in the future.

Meaning that even when a persistent database is used, the settings will not be persistent through closing and reopening of the database.

**Comment by parvizfarnia** [comment]
Thank you for clarifying this.

> Meaning that even when a persistent database is used, the settings will not be persistent through closing and reopening of the database.

But at least during each process execution they remain persistent (only for the current running process, I mean), right?

**Reply by Tishj** [reply]
Yes that's correct

**Comment by parvizfarnia** [comment]
Thank you very much for your time and your help.
